<img src="ampir_hex.png" width="90" align="right" height="100" />

**ampir** (short for **a**nti**m**icrobial **p**eptide prediction
**i**n **r** ) identifies antimicrobial peptides (AMPs) based on physico-chemical properties of amino acid sequences. It is optimised for scanning large collections of sequences (eg the predicted proteome for an organism) to identify a suite of putative AMPs.

**ampir** was developed by [Legana Fingerhut](https://twitter.com/leganafingerhut) and [Ira Cooke](https://twitter.com/iracooke) in the [Marine Omics](https://www.marine-omics.net/) and [Marine Molecular Biology](https://www.marine-molecular-biology.group/) labs at [James Cook University](https://www.jcu.edu.au/ctbmb).  It is [open source](https://github.com/legana/ampir) and available for use via this web page or as an [R package](https://cran.r-project.org/web/packages/ampir/index.html).

If you use **ampir** in your research please cite the following paper

> Fingerhut, L.C.H.W., Miller, D.J., Strugnell, J.M., Daly, N.L., Cooke, I.R., 2020. ampir: an R package for fast 
> genome-wide prediction of antimicrobial peptides. Bioinformatics. https://doi.org/10.1093/bioinformatics/btaa653


